content/asm_6.md

```diff
@@ -1,181 +1,561 @@
+        syscall
+        ;; Exit from the program
+        jmp _exit
+
+;; Exit from the program.
+_exit:
+    ;; Specify the number of the system call (60 is `sys_exit`).
+    mov rax, SYS_EXIT
+    ;; Set the first argument of `sys_exit` to 0. The 0 status code is success.
+    mov rdi, EXIT_CODE
+    ;; Call the `sys_exit` system call.
+    syscall
+```
+
+If both vectors have the same number of components, we can proceed to the calculation. To do that, we store the addresses of both vectors in the `rdi` and `rsi` registers, and put the number of components within the vectors into the `rdx` register. The `_dot_product` function does the main job. Let's take a look at the code of this function:
 
 ```assembly
-lcall $section, $offset
+;; Calculate the dot product of the two vectors.
+_dot_product:
+        ;; Reset the value of the rax register to 0.
+        xor rax, rax
+        ;; Reset the value of the xmm1 register to 0.
+        pxor xmm1, xmm1
+        ;; Current rdx contains the number of floating-point values within the vectors.
+        ;; Multiply it by 8 using shift left operation to get the number of bytes 
+        ;; occupied by these values.
+        sal rdx, 3
+;; Calculate the the dot product in the loop.
+_loop:
+        ;; Move the floating-point value from the first vector to the xmm0 register.
+        movsd xmm0, [rdi + rax]
+        ;; Multiply the floating-point value from the second vector by the value from the first vector
+        ;; and store the result in the xmm0 register.
+        mulsd xmm0, [rsi + rax]
+        ;; Move to the next floating-point values in the vector buffers.
+        add rax, 8
+        ;; Add the result of multiplying the floating-point values from the two vectors into the xmm1 register.
+        addsd xmm1, xmm0
+        ;; Check if we went through all the floating-point values in the vector buffers.
+        cmp rax, rdx
+        ;; If not, repeat the loop.
+        jne _loop
+        ;; Move the result to the xmm0 register.
+        movapd xmm0, xmm1
+        ;; Return from the `_dot_product` back to the `_calculate_dot_product`.
+        ret
 ```
 
-Far jump - a jump to an instruction located in a different segment than the current code segment but at the same privilege level, sometimes referred to as an intersegment jump.
+At the beginning of the function, we prepare the `rax` and `xmm1` registers by resetting them to zero. The `rax` register will contain the offset within the vector buffers, and the `xmm1` register will accumulate the result of our program. The number of floating-point values within the vectors is stored in the `rdx` register. Since we will move the pointer within the buffers, we need to know how many bytes occupied by these values. To do this we multiple the value of the `rdx` register by `8`. To do this we use the `sal` instruction which executes [arithmetic left shift](https://en.wikipedia.org/wiki/Arithmetic_shift) on the given operand.
```

Comment:
```suggestion
At the beginning of the function, we prepare the `rax` and `xmm1` registers by resetting them to zero. The `rax` register will contain the offset within the vector buffers, and the `xmm1` register will accumulate the result of our program. The number of floating-point values within the vectors is stored in the `rdx` register. Since we will move the pointer within the buffers, we need to know how many bytes are occupied by these values. To do this, we multiply the value of the `rdx` register by `8` using the `sal` instruction, which executes [arithmetic left shift](https://en.wikipedia.org/wiki/Arithmetic_shift) on the given operand.
```